### PR TITLE
[BUG] Preserve float32 dtype in affected transformers.

### PR DESCRIPTION
**Pull request by 4nmus**
#### Reference Issues/PRs

Fixes part of #3723.

#### What does this implement/fix? Explain your changes.

Preserves `float32` output precision in the following transformers:
DWTTransformer
Resizer
SlopeTransformer
SupervisedIntervals
PLASeriesTransformer

Previously, these transformers could unnecessarily promote float32 input to float64 through explicit casts or default NumPy allocations.
The changes preserve the following dtype behavior:
-float32 input produces float32 output
-float64 input produces float64 output
-integer input is promoted to float64

Regression tests were added or updated to verify dtype behavior for float32, float64, and integer inputs.

ClaSPTransformer, which is mentioned in #3723, is not included in this PR because its existing float64 conversion is related to a separate Numba typing constraint.

#### Does your contribution introduce a new dependency? If yes, which one?

No.

#### Any other comments?

ClaSPTransformer still returns float64 since it forces the input to be float64 due to Numba. 

##### For all contributions
- [X] I've added myself to the [list of contributors](https://github.com/aeon-toolkit/aeon/blob/main/.all-contributorsrc). Alternatively, you can use the [@all-contributors](https://allcontributors.org/docs/en/bot/usage) bot to do this for you **after** the PR has been merged.
- [X] The PR title starts with either [ENH], [MNT], [DOC], [BUG], [REF], [DEP] or [GOV] indicating whether the PR topic is related to enhancement, maintenance, documentation, bugs, refactoring, deprecation or governance.

##### For new estimators and functions
- [ ] I've added the estimator/function to the online [API documentation](https://www.aeon-toolkit.org/en/latest/api_reference.html).
- [ ] (OPTIONAL) I've added myself as a `__maintainer__` at the top of relevant files and want to be contacted regarding its maintenance. Unmaintained files may be removed. This is for the full file, and you should not add yourself if you are just making minor changes or do not want to help maintain its contents.

##### For developers with write access
- [ ] (OPTIONAL) I've updated aeon's [CODEOWNERS](https://github.com/aeon-toolkit/aeon/blob/main/CODEOWNERS) to receive notifications about future changes to these files.

### CODE REVIEW

File: aeon/transformations/collection/interval_based/tests/test_intervals.py [hunk 1]
```diff
@@ -56,3 +59,39 @@ def test_supervised_transformers():
...
 
     assert X_t.shape == (X.shape[0], 8)
+
+
+@pytest.mark.parametrize(
+    "dtype, expected_dtype",
+    [
+        (np.float32, np.float32),
+        (np.float64, np.float64),
+        (np.int64, np.float64),
```
Comment (baraline) [inline aeon/transformations/collection/interval_based/tests/test_intervals.py]: need to also add int32, float64

File: aeon/transformations/collection/interval_based/_supervised_intervals.py [hunk 2]
```diff
@@ -208,7 +208,8 @@ def _fit_transform(self, X, y=None):
         for i in range(1, self.n_intervals):
             Xt = np.hstack((Xt, transformed_intervals[i]))
 
-        return Xt
+        out_dtype = np.float32 if X.dtype == np.float32 else np.float64
+        return Xt.astype(dtype=out_dtype, copy=False)
```
Comment (baraline) [inline aeon/transformations/collection/interval_based/_supervised_intervals.py:212]: We shouldn't neet to cast the input here, it must natively used float32 (or 64) in the generate intervals function.

What we want is that if we have a float 32 input, computation are made on float 32 to benefit from the reduced memory footprint (and computation time to some extent)

File: aeon/transformations/collection/interval_based/_supervised_intervals.py [hunk 3]
```diff
@@ -251,7 +252,8 @@ def _transform(self, X, y=None):
             for i in range(len(self.intervals_))
         )
 
-        Xt = np.zeros((X.shape[0], len(transform)))
+        out_dtype = np.float32 if X.dtype == np.float32 else np.float64
+        Xt = np.zeros((X.shape[0], len(transform)), dtype=out_dtype)
```
Comment (baraline) [inline aeon/transformations/collection/interval_based/_supervised_intervals.py]: Same comment as above, we want the interval computation to use 32 bit precision, not simply cast the output.

So `_transform_intervals` itself should return the appropriate dtype already.

File: aeon/transformations/collection/tests/test_dwt.py [hunk 4]
```diff
@@ -111,3 +111,20 @@ def test_dwt_performs_correcly_along_each_dim():
...
     )
     np.testing.assert_array_almost_equal(res, orig)
+
+
+@pytest.mark.parametrize(
+    "dtype, expected_dtype",
+    [
+        (np.float32, np.float32),
+        (np.float64, np.float64),
+        (np.int64, np.float64),
```
Comment (baraline) [inline aeon/transformations/collection/tests/test_dwt.py]: similar to before, add int32, float64

File: aeon/transformations/collection/tests/test_dwt.py [hunk 5]
```diff
@@ -111,3 +111,20 @@ def test_dwt_performs_correcly_along_each_dim():
...
+    ],
+)
+def test_dwt_preserves_float_precision(dtype, expected_dtype):
+    """Check dwt preserves float32 and promotes proper dtype output."""
+    X = np.arange(16, dtype=dtype).reshape(2, 1, 8)
```
Comment (baraline) [inline aeon/transformations/collection/tests/test_dwt.py]: use the testing utils functions to generate testing data, as you did in the inerval tests (make_example_3d_numpy)

File: aeon/transformations/collection/tests/test_slope_transformer.py [hunk 6]
```diff
@@ -75,3 +75,23 @@ def test_slope_performs_correcly_along_each_dim():
...
     )
     np.testing.assert_array_almost_equal(res, orig, decimal=5)
+
+
+@pytest.mark.parametrize(
+    "dtype, expected_dtype",
+    [
+        (np.float32, np.float32),
+        (np.float64, np.float64),
+        (np.int64, np.float64),
```
Comment (baraline) [inline aeon/transformations/collection/tests/test_slope_transformer.py]: also int32, float64

File: aeon/transformations/collection/unequal_length/tests/test_resize.py [hunk 7]
```diff
@@ -87,3 +87,23 @@ def test_incorrect_arguments():
...
     with pytest.raises(ValueError, match="resized_length must be"):
         resizer.fit_transform(X)
+
+
+@pytest.mark.parametrize(
+    "dtype, expected_dtype",
+    [
+        (np.float32, np.float32),
+        (np.float64, np.float64),
+        (np.int64, np.float64),
```
Comment (baraline) [inline aeon/transformations/collection/unequal_length/tests/test_resize.py]: also int32,float64

File: aeon/transformations/collection/tests/test_slope_transformer.py [hunk 8]
```diff
@@ -75,3 +75,23 @@ def test_slope_performs_correcly_along_each_dim():
...
+    ],
+)
+def test_slope_preserves_float_precision(dtype, expected_dtype):
+    """Check that Slope preserved float32 and promotes proper dtype output."""
+    X = np.array(
+        [[[4, 6, 10, 12, 8, 6, 5, 5]]],
+        dtype=dtype,
+    )
```
Comment (baraline) [inline aeon/transformations/collection/tests/test_slope_transformer.py]: again, use the testing utils to generate data

File: aeon/transformations/collection/unequal_length/tests/test_resize.py [hunk 9]
```diff
@@ -87,3 +87,23 @@ def test_incorrect_arguments():
...
+    ],
+)
+def test_resizer_preserves_float_precision(dtype, expected_dtype):
+    """Test Resizer preserved float32 and promotes proper dtype output."""
+    X = [
+        np.array([[0, 1, 2, 3]], dtype=dtype),
+        np.array([[0, 2, 4, 6, 8]], dtype=dtype),
```
Comment (baraline) [inline aeon/transformations/collection/unequal_length/tests/test_resize.py:100]: I don't think the current testing util to generate variable length list allows dtype args, so this one is fine for now.

**Review by baraline — CHANGES_REQUESTED** [review]
Thanks for working on this, the issue is a bit more complex than simply casting the output, we want the internal computation to be performed in the appropriate precision, to benefit from the reduced memory footprint.

I didn't comment on _dwt, _slope and _pla as the issues are the same that those pointed out in intervals.

If you need any pointers don't hesitate to ask questions.

**Comment by 4nmus** [comment]
Thank you for the review! I updated implementation so that computations itself preserve float32 within in `_pla`, `_supervised_intervals` and `_slope`. However, there is an issue within _supervised_intervals:` row_*` feature funcitons force float64 regardless of changes, it's already covered in #3724 / #3749  so I kept the SupervisedIntervals working arrays and interval buffers in X.dtype and just normalized feature outputs at that boundary. I also added fixedd and added int32 in tests. I hope that is what was missing.

File: aeon/transformations/collection/interval_based/tests/test_intervals.py (hunk 1 above)
Comment (4nmus) [reply aeon/transformations/collection/interval_based/tests/test_intervals.py]: Added

File: aeon/transformations/collection/interval_based/_supervised_intervals.py (hunk 3 above)
Comment (4nmus) [reply aeon/transformations/collection/interval_based/_supervised_intervals.py]: _transform_intervals now returns values in the working X.dtype, and _transform allocates its result directly with that dtype instead of correcting the dtype

File: aeon/transformations/collection/interval_based/_supervised_intervals.py (hunk 2 above)
Comment (4nmus) [reply aeon/transformations/collection/interval_based/_supervised_intervals.py:212]: I removed final cast. The input is promoted once and interval  buffers now use X.dtype, so float32 is preserved throughout the SupervisedIntervals.

File: aeon/transformations/collection/tests/test_dwt.py (hunk 4 above)
Comment (4nmus) [reply aeon/transformations/collection/tests/test_dwt.py]: added

File: aeon/transformations/collection/tests/test_slope_transformer.py (hunk 8 above)
Comment (4nmus) [reply aeon/transformations/collection/tests/test_slope_transformer.py]: Updated to use make_example_3d_numpy

File: aeon/transformations/collection/tests/test_slope_transformer.py (hunk 6 above)
Comment (4nmus) [reply aeon/transformations/collection/tests/test_slope_transformer.py]: added

File: aeon/transformations/collection/tests/test_dwt.py (hunk 5 above)
Comment (4nmus) [reply aeon/transformations/collection/tests/test_dwt.py]: now uses  make_example_3d_numpy

**Comment by baraline** [comment]
Great ! Overall looks fine but I'll try to have a closer look at it this week. 

Thanks for continuing to work on this.

**Comment by baraline** [comment]
@MatthewMiddlehurst, do you think it would be worth having a internal estimator attribute (would tag system be appropriate to declare the capability ?) with the input precision (32/64) set during fit and reused across all the estimator declarations ?
It would avoid using the same dtype everytime and would make a clean standard across all estimators to declare and use the precision.

We can set it during input preprocessing similarly to how @4nmus wrote it in the test :
```
internal_dtype = np.float32 if input_dtype == np.float32 else dtype = np.float64
```

**Comment by MatthewMiddlehurst** [comment]
Yes could be useful. Lots of possible ways to handle dtypes which we haven't really done anything with really.

File: aeon/transformations/collection/interval_based/_supervised_intervals.py [hunk 10]
```diff
@@ -241,6 +241,8 @@ def _fit(self, X, y=None):
         return self
 
     def _transform(self, X, y=None):
+        X = X / 1
+
```
Comment (MatthewMiddlehurst) [inline aeon/transformations/collection/interval_based/_supervised_intervals.py:245]: why?

File: aeon/transformations/collection/_slope.py [hunk 11]
```diff
@@ -62,17 +60,20 @@ def _transform(self, X, y=None):
         # Get information about the dataframe
         n_cases, n_channels, n_timepoints = X.shape
         self._check_parameters(n_timepoints)
-        full_data = []
+
+        X = X / 1
```
Comment (MatthewMiddlehurst) [inline aeon/transformations/collection/_slope.py:64]: same here (and everywhere else)

File: aeon/transformations/collection/interval_based/_supervised_intervals.py (hunk 10 above)
Comment (4nmus) [reply aeon/transformations/collection/interval_based/_supervised_intervals.py:245]: This would establish proper dtype before computation. Float32 and float64 is preserved while int input promoted to float64 for further computation using numpy's promotion rules.It was used in #3707 like return X / 1 to preserve the dtype so i followed same approach as in `_as_normalized_float` in #3707. It does not cast the final result back to requested dtype. I can also change it to a helper function but it's seems to be a cleaner version.

Comment (MatthewMiddlehurst) [reply aeon/transformations/collection/interval_based/_supervised_intervals.py:245]: not really sure I like the X/1 thing for both really, seems a bit hacky.

Comment (4nmus) [reply aeon/transformations/collection/interval_based/_supervised_intervals.py:245]: Okay, i can do it with `dtype = np.result_type(X.dtype, np.float32)`. So it's going to be something like that:
```
dtype = np.result_type(X.dtype, np.float32)
Xt = np.empty((n_cases, n_channels, n_intervals), dtype=dtype)
Y = np.arange(interval_size, dtype=dtype)
```
Is this okay?
